File: app/src/main/java/su/linka/pictures/activity/BroadcastReceiverActivity.java
package su.linka.pictures.activity;

import static androidx.activity.result.ActivityResultCallerKt.registerForActivityResult;

import android.app.Activity;
import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;

import androidx.annotation.Nullable;

import com.google.firebase.analytics.FirebaseAnalytics;

import java.io.File;
import java.io.IOException;

import su.linka.pictures.AnalyticsEvents;
import su.linka.pictures.Callback;
import su.linka.pictures.R;
import su.linka.pictures.SetsManager;
import su.linka.pictures.Utils;
import su.linka.pictures.components.ConfirmDialog;

public class BroadcastReceiverActivity extends Activity {
    private FirebaseAnalytics firebaseAnalytics;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        firebaseAnalytics = FirebaseAnalytics.getInstance(this);
//        setContentView(R.layout.activity_main);
        final Intent intent = getIntent();
        final String action = intent.getAction();

        if (Intent.ACTION_VIEW.equals(action)) {
            //uri = intent.getStringExtra("URI");
            Uri uri = intent.getData();
            ConfirmDialog
                    .showConfirmDialog(this, R.string.add_set_to_library, new Callback() {
                        @Override
                        public void onDone(Object result) {
                            copy(uri);
                        }


                        @Override
                        public void onFail(Exception error) {
finish();
                        }
                    });
            // now you call whatever function your app uses
            // to consume the txt file whose location you now know
        } else {
        }
    }

    private void copy(Uri uri) {
        try {
            Utils.copy( getContentResolver().openInputStream(uri), new File ((new SetsManager(this)).getSetsDirectory(),uri.getPathSegments().get(uri.getPathSegments().size()-1) ));
        } catch (IOException e) {
            e.printStackTrace();
            finish();
        }
        firebaseAnalytics.logEvent(AnalyticsEvents.ADD_SET, null);

        Intent intent = new Intent(this, MainActivity.class);
        startActivity(intent);
        finish();
    }
}

File: app/src/main/java/su/linka/pictures/activity/BroadcastReceiverActivity.kt
package su.linka.pictures.activity

import android.app.Activity
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import com.google.firebase.analytics.FirebaseAnalytics
import su.linka.pictures.AnalyticsEvents
import su.linka.pictures.Callback
import su.linka.pictures.R
import su.linka.pictures.SetsManager
import su.linka.pictures.Utils
import su.linka.pictures.components.ConfirmDialog
import java.io.File
import java.io.IOException

class BroadcastReceiverActivity : Activity() {

    private lateinit var analytics: FirebaseAnalytics

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        analytics = FirebaseAnalytics.getInstance(this)
        val action = intent?.action
        if (Intent.ACTION_VIEW == action) {
            val uri = intent.data
            if (uri != null) {
                askCopy(uri)
            } else {
                finish()
            }
        } else {
            finish()
        }
    }

    private fun askCopy(uri: Uri) {
        ConfirmDialog.showConfirmDialog(this, R.string.add_set_to_library, object : Callback<Any?>() {
            override fun onDone(result: Any?) {
                copy(uri)
            }

            override fun onFail(error: Exception?) {
                finish()
            }
        })
    }

    private fun copy(uri: Uri) {
        val setsManager = SetsManager(this)
        val fileName = uri.lastPathSegment ?: run {
            finish()
            return
        }
        val destination = File(setsManager.getSetsDirectory(), fileName)
        try {
            contentResolver.openInputStream(uri)?.use { input ->
                Utils.copy(input, destination)
            } ?: throw IOException("Unable to open input stream")
        } catch (error: IOException) {
            error.printStackTrace()
            finish()
            return
        }
        analytics.logEvent(AnalyticsEvents.ADD_SET, null)
        startActivity(Intent(this, MainActivity::class.java).apply {
            addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP or Intent.FLAG_ACTIVITY_NEW_TASK)
        })
        finish()
    }
}
